File: Archived/DP/10844.java
import java.util.*;
import java.io.*;

public class Main{
    static BufferedReader br;
    static BufferedWriter bw;
    static int[][] cache;
    static int N;

    public static void main(String[] args) throws IOException{
        br = new BufferedReader(new InputStreamReader(System.in));
        bw = new BufferedWriter(new OutputStreamWriter(System.out));
        N = Integer.parseInt(br.readLine());
        cache = new int[N][10];
        for(int i=0;i<N;i++)
            Arrays.fill(cache[i],-1);
        int ans=0;
        for(int i=1;i<10;i++)
            ans = (ans%1000000000 + recur(0,i)%1000000000)%1000000000;
        bw.write(Integer.toString(ans)+"\n");
        bw.flush();bw.close();br.close();
    }

    static int recur(int idx, int curr){
        if(curr<0 || curr>9)
            return 0;
        if(idx==N-1)
            return 1;
        if(cache[idx][curr]!= -1)
            return cache[idx][curr];
        
        int tmp = (recur(idx+1,curr+1)%1000000000 + recur(idx+1,curr-1)%1000000000)%1000000000;
        cache[idx][curr] = tmp;
        return tmp;
    }
}

File: Archived/14단계/10844.py
import sys

def recur(y,x):
    global cache
    if x < 0 or x > 9:
        return 0
    
    if y == 1:
        return 1
    
    if x == 0:
        return recur(y-1,x+1)
    
    if x == 9:
        return recur(y-1,x-1)
    
    
    if cache[y][x] != None:
        return cache[y][x]
    
    tmp = (recur(y-1,x-1)%1_000_000_000 + recur(y-1,x+1)%1_000_000_000)%1_000_000_000
    cache[y][x] = tmp
    
    return tmp

N = int(sys.stdin.readline().strip())
cache = [[None for i in range(9)] for j in range(N+1)]
cache[1] = [1 for i in range(9)]
ans = 0
for i in range(9):
    ans += recur(N,i)

# 마지막 출력단계에서도 꼭 나머지 연산을 취해주자..
print(ans%1_000_000_000)
